rules/consistent-indexof-check.js

```diff
@@ -0,0 +1,116 @@
+'use strict';
+const evaluateLiteralUnaryExpression = require('./utils/evaluate-literal-unaryexpression.js');
+const resolveVariableName = require('./utils/resolve-variable-name.js');
+
+const MESSAGE_ID = 'consistent-indexof-check';
+const messages = {
+	[MESSAGE_ID]: 'Prefer `{{replacement}}` over `{{value}}`.',
+};
+
+const comparisonMap = {
+	'<': '>',
+	'<=': '>=',
+	'>': '<',
+	'>=': '<=',
+};
+
+/**
+Check if the node is a call expression of `indexOf` or `lastIndexOf` method.
+
+@param {import('estree').Node} node
+@returns {node is import('estree').CallExpression}
+*/
+function isIndexOfCallExpression(node) {
+	return (
+		node?.type === 'CallExpression'
+		&& node.callee?.type === 'MemberExpression'
+		&& ['indexOf', 'lastIndexOf', 'findIndex', 'findLastIndex'].includes(node.callee.property.name)
+		&& node.arguments.length === 1
+	);
+}
+
+/**
+Determine the appropriate replacement based on the operator and value.
+
+@param {string} operator
+@param {number} value
+@param {string} variableName
+@returns {string | undefined}
+*/
+function getReplacement(operator, value, variableName) {
+	if ((operator === '<' && value <= 0) || (operator === '<=' && value <= -1)) {
+		return `${variableName} === -1`;
+	}
+
+	if ((operator === '>' && value === -1) || (operator === '>=' && value === 0)) {
+		return `${variableName} !== -1`;
+	}
+}
+
+/**
+Check if the node is a number literal or a unary expression resolving to a number.
+
+@param {import('estree').Node} node
+@returns {node is import('estree').Literal}
+*/
+function isNumberLiteral(node) {
+	if (node.type === 'UnaryExpression' && ['-', '+', '~'].includes(node.operator)) {
+		return isNumberLiteral(node.argument);
+	}
+
+	return node.type === 'Literal' && typeof node.value === 'number';
+}
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => ({
+	/** @param {import('estree').BinaryExpression} node */
+	BinaryExpression(node) {
+		const [identifier, literal, operator]
+			= [node.left, node.right].some(n => isNumberLiteral(n)) && [node.left, node.right].some(n => n.type === 'Identifier')
+				? (node.left.type === 'Identifier'
+					? [node.left.name, evaluateLiteralUnaryExpression(node.right), node.operator] // Index === -1
+					: [node.right.name, evaluateLiteralUnaryExpression(node.left), comparisonMap[node.operator]]) // -1 === index
+				: [];
+
+		if (!identifier) {
+			return;
+		}
+
+		const replacement = getReplacement(operator, literal, identifier);
+
+		if (!replacement) {
+			return;
+		}
+
+		const variableFound = resolveVariableName(identifier, context.sourceCode.getScope(node));
```

Comment:
What do you think listen on assignment `const index = ...indexOf(..)` then look for comparison instead?

Reply:
This implementation may get a better performance. (Because it searches for `const index = ...indexOf(..)` statement first, and then searches for `BinaryExpression`)

But it is also more complicated, There are countless possibilities for `BinaryExpression` to appear, and traverse almost the entire AST, I think it is not worth it.

for example:

```js
const index = "foo".indexOf("o");

function test() {
	class Man {
		method() {
			return {
				nest () {
					if (index > -1) { // ❌ You have to traverse from the top level to here
					
					}
				},
				noop () {
					// ❌ Even if there is no node here, you must traverse here to check
				}
			}
		}
	}
}
```

Reply:
It's easy to find them, since `index` variable already include all the references.

Reply:
All right, I will try it later

Reply:
Done in 732ae0560cd55526c8e77c45ef3a7a479bae4d9a

Reply:
Should we catch the case of `AssignmentExpression` ?

```
let index = 0;

// do something

index = foo.indexOf("xx")

// ...

if (index > -1) {

}
```

Reply:
> Should we catch the case of `AssignmentExpression` ?
> 
> ```js
> let index = 0;
> 
> // do something
> 
> index = foo.indexOf("xx")
> 
> // ...
> 
> if (index > -1) {
> 
> }
> ```

Implemented

Reply:
> Should we catch the case of AssignmentExpression ?

No, should only check  work for declaration with `const`

Reply:
> No, should only check declaration with const

Is there any reason for this?

Reply:
We can't know how it will be used, we don't want false positives.